src/Microsoft.Framework.Caching.SqlServer/DatabaseOperations.cs

```diff
@@ -0,0 +1,377 @@
+                if (!reader.Read())
+                {
+                    throw new InvalidOperationException(
+                        $"Could not retrieve information of table with schema '{SchemaName}' and " +
+                        $"name '{TableName}'. Make sure you have the table setup and try again. " +
+                        $"Connection string: {ConnectionString}");
+                }
+            }
+        }
+
+        public virtual async Task GetTableSchemaAsync()
+        {
+            using (var connection = new SqlConnection(ConnectionString))
+            {
+                var command = new SqlCommand(SqlQueries.TableInfo, connection);
+                await connection.OpenAsync();
+
+                var reader = await command.ExecuteReaderAsync(CommandBehavior.SingleRow);
+
+                if (!await reader.ReadAsync())
+                {
+                    throw new InvalidOperationException(
+                        $"Could not retrieve information of table with schema '{SchemaName}' and " +
+                        $"name '{TableName}'. Make sure you have the table setup and try again. " +
+                        $"Connection string: {ConnectionString}");
+                }
+            }
+        }
+
+        public void DeleteExpiredCacheItems()
+        {
+            var utcNowDateTime = SystemClock.UtcNow.UtcDateTime;
+
+            using (var connection = new SqlConnection(ConnectionString))
+            {
+                var command = new SqlCommand(SqlQueries.DeleteExpiredCacheItems, connection);
+                command.Parameters.AddWithValue("UtcNow", SqlDbType.DateTime, utcNowDateTime);
+
+                connection.Open();
+
+                var effectedRowCount = command.ExecuteNonQuery();
+            }
+        }
+
+        public virtual void SetCacheItem(string key, byte[] value, DistributedCacheEntryOptions options)
+        {
+            var utcNowDateTime = SystemClock.UtcNow.UtcDateTime;
+
+            var expirationInfo = CacheItemExpiration.GetExpirationInfo(utcNowDateTime, options);
+
+            using (var connection = new SqlConnection(ConnectionString))
+            {
+                var upsertCommand = new SqlCommand(SqlQueries.SetCacheItem, connection);
+                upsertCommand.Parameters
+                    .AddCacheItemId(key)
+                    .AddCacheItemValue(value)
+                    .AddExpiresAtTime(expirationInfo.ExpiresAtTimeUTC)
+                    .AddSlidingExpirationInTicks(expirationInfo.SlidingExpiration)
+                    .AddAbsoluteExpiration(expirationInfo.AbsoluteExpirationUTC);
+
+                connection.Open();
+
+                try
+                {
+                    upsertCommand.ExecuteNonQuery();
+                }
+                catch (SqlException)
+                {
+                    // There is a possibility that multiple requests can try to add the same item to the cache, in
```

Comment:
Is that the only possible error? If so log it as Info or Warning.  If there are other possible errors we shouldn't be masking them.